INT-4174: Expose Channel Subscriber Count

garyrussell opened the pull request:
JIRA: https://jira.spring.io/browse/INT-4174

I considered just using the channel counter but I suppose there is a (small) possibility
that someone might possibly use a custom dispatcher and subscribe to it directly.


__cherry-pick to 4.3.x__

garyrussell committed:
INT-4174: Expose Channel Subscriber Count

JIRA: https://jira.spring.io/browse/INT-4174

I considered just using the channel counter but I suppose there is a (small) possibility
that someone might possibly use a custom dispatcher and subscribe to it directly.

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/channel/AbstractSubscribableChannel.java:
```diff
@@ -37,19 +38,30 @@
+		if (dispatcher instanceof AbstractDispatcher) {
+			return ((AbstractDispatcher) dispatcher).getHandlerCount();
+		}
+		return this.handlerCounter.get();
```
Good idea - working on it now...

artembilan replied:
Great! 
But I think we can't backport it to `4.3.x`. 
Therefore that work should be done in the separate PR.

What is in your mind ?

garyrussell replied:
Just cherry-pick the first commit.

Let me know if you prefer a separate PR.

garyrussell committed:
Polishing -- ** master only **

Eliminate conditional logic in AbstractMessageChannel by adding getHandlerCount() to
MessageDispatcher.

Do not cherry pick.

artembilan reviewed (approved):
Merging and cherry-picking to `4.3.x` only first commit...

artembilan commented:
Merged as https://github.com/spring-projects/spring-integration/commit/96ff547c8c06c9e465270485890974ccfbdfdebb and cherry-picked to `4.3.x` only first commit.

Also note the fix for JPA tests race condition.
